crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -542,12 +544,60 @@ async fn execute_transactions(
     authority_state
         .database
         .epoch_store()
-        .insert_pending_certificates(&txns)?;
-
-    authority_state.transaction_manager.enqueue(txns).await?;
+        .insert_pending_certificates(&synced_txns)?;
 
     authority_state
+        .transaction_manager
+        .enqueue(synced_txns)
+        .await?;
+
+    let effects = authority_state
+        .database
+        .notify_read_effects(synced_tx_digests)
+        .await?;
+
+    assert_eq!(
+        effects.len(),
+        num_synced_txns,
+        "Missing txn effects from transaction manager"
+    );
+    Ok(())
+}
+
+fn get_transactions(
+    digests: &Vec<TransactionDigest>,
+    authority_state: Arc<AuthorityState>,
+) -> SuiResult<Vec<VerifiedCertificate>> {
+    let synced_txns: Vec<VerifiedCertificate> = authority_state
+        .database
+        .perpetual_tables
+        .synced_transactions
+        .multi_get(digests)?
+        .into_iter()
+        .flatten()
+        .map(|tx| tx.into())
+        .collect();
+
+    // Get set of digests that did not exist in synced_transactions
+    let synced_digests: HashSet<TransactionDigest> =
+        HashSet::from_iter(synced_txns.clone().into_iter().map(|tx| *tx.digest()));
+    let diff: Vec<TransactionDigest> = HashSet::from_iter(digests.clone().into_iter())
+        .difference(&synced_digests)
+        .collect::<HashSet<&TransactionDigest>>()
+        .into_iter()
+        .copied()
+        .collect();
+
+    // Ensure that effects exist for any transaction of this checkpoint that
+    // are not in synced transactions table, as this suggests that they
+    // were executed locally
+    let _ = authority_state
         .database
-        .notify_read_effects(tx_digests)
-        .await
+        .perpetual_tables
+        .executed_effects
```

Comment:
1. Calling `effects_exists()` seems more suitable.
2. My worry is that if data sync skips writing to synced_certificates when the certificate exists in pending_certificates, the certificate that does not exist in synced_certificates may still be executing, and this condition may not always be correct. We can either do the look up in pending_certificates table here as well, or skip this check and rely on calling notify_read_effects() on all tx_digests.

Reply:
True, probably just awaiting notify for all the effects for all txns in the input would be the most fail-safe way. Will fix in the morning